File: app/schemas/skill.py
import uuid

from pydantic import BaseModel, ConfigDict, Field

from app.models.exercise import TargetStat


class SkillRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str


class SkillCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)


class SkillUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)


class SkillStatWeightRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    skill_id: uuid.UUID
    stat_type: TargetStat
    weight: float = Field(ge=0.0, le=1.0)


class SkillStatWeightCreate(BaseModel):
    stat_type: TargetStat
    weight: float = Field(ge=0.0, le=1.0)


class SkillStatWeightUpdate(BaseModel):
    stat_type: TargetStat | None = None
    weight: float | None = Field(default=None, ge=0.0, le=1.0)


class SkillTagRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    skill_id: uuid.UUID
    exercise_id: uuid.UUID
    transfer_note: str


class SkillTagCreate(BaseModel):
    exercise_id: uuid.UUID
    transfer_note: str = Field(min_length=1)


class SkillTagUpdate(BaseModel):
    transfer_note: str | None = Field(default=None, min_length=1)


class SkillMilestoneRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    skill_id: uuid.UUID
    threshold: int = Field(ge=0, le=100)
    title: str
    description: str


class SkillMilestoneCreate(BaseModel):
    threshold: int = Field(ge=0, le=100)
    title: str = Field(min_length=1, max_length=255)
    description: str = Field(min_length=1)


class SkillMilestoneUpdate(BaseModel):
    threshold: int | None = Field(default=None, ge=0, le=100)
    title: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = Field(default=None, min_length=1)


class NextMilestoneRead(BaseModel):
    title: str
    threshold: int
    points_remaining: float


class SkillSummaryRead(BaseModel):
    id: uuid.UUID
    name: str
    value: float
    next_milestone: NextMilestoneRead | None


class StatContributionRead(BaseModel):
    stat_type: TargetStat
    weight: float
    effective_value: float
    contribution: float


class SkillMilestoneStatusRead(BaseModel):
    id: uuid.UUID
    threshold: int
    title: str
    description: str
    achieved: bool


class SkillDetailRead(BaseModel):
    id: uuid.UUID
    name: str
    value: float
    stat_breakdown: list[StatContributionRead]
    milestones: list[SkillMilestoneStatusRead]

File: frontend/src/types/skill.ts
import type { TargetStat } from './exercise'

export interface SkillOption {
  id: string
  name: string
  required_level: number
}

export interface UserSkillPreference {
  skill_id: string
  name: string
}

export interface NextMilestoneRead {
  title: string
  threshold: number
  points_remaining: number
}

export interface SkillSummaryRead {
  id: string
  name: string
  value: number
  next_milestone: NextMilestoneRead | null
  required_level: number
}

export interface StatContributionRead {
  stat_type: TargetStat
  weight: number
  effective_value: number
  contribution: number
}

export interface SkillMilestoneStatusRead {
  id: string
  threshold: number
  title: string
  description: string
  achieved: boolean
}

export interface SkillDetailRead {
  id: string
  name: string
  value: number
  stat_breakdown: StatContributionRead[]
  milestones: SkillMilestoneStatusRead[]
}

// -- admin CRUD (raw records, no computed value/progress) --

export interface SkillWrite {
  name: string
  required_level: number
}

export interface SkillStatWeightRead {
  id: string
  skill_id: string
  stat_type: TargetStat
  weight: number
}

export interface SkillStatWeightWrite {
  stat_type: TargetStat
  weight: number
}

export interface SkillTagRead {
  id: string
  skill_id: string
  exercise_id: string
  transfer_note: string
}

export interface SkillTagWrite {
  exercise_id: string
  transfer_note: string
}

// Distinct from SkillMilestoneStatusRead above -- that one is the
// user-progress view (with `achieved`), this is the raw admin record.
export interface SkillMilestoneRead {
  id: string
  skill_id: string
  threshold: number
  title: string
  description: string
}

export interface SkillMilestoneWrite {
  threshold: number
  title: string
  description: string
}
